### PR TITLE
8281315: Unicode, (?i) flag and backreference throwing IndexOutOfBounds Exception

### PR DESCRIPTION
**Pull request by igraves**
This is a fix in the buggy way CIBackRef traverses unicode characters that could be variable-length. Originally it followed the approach that BackRef does, but failed to account for unicode characters that could be 2 chars-long. The upper bound (groupSize) for the traversing loop is set by the difference between group start and stop indexes. This works for single char characters and it also works for case-sensitive comparisons because byte-by-byte comparisons are acceptable, but it doesn't work for a comparison where some kind of normalization (i.e. case) is required. This fix adjusts the upper bound for the loop that traverses the character when a two-char character is encountered.

An alternative was to check the length of the group size by scanning the group in advance and converting to code points, but this could potentially result in multiple scans and codepoint conversions of the same matcher group which could be long. The solution that adjusts the loop bounds on the fly avoids this case.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281315](https://bugs.openjdk.java.net/browse/JDK-8281315): Unicode, (?i) flag and backreference throwing IndexOutOfBounds Exception

### Reviewers
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7501/head:pull/7501` \
`$ git checkout pull/7501`

Update a local copy of the PR: \
`$ git checkout pull/7501` \
`$ git pull https://git.openjdk.java.net/jdk pull/7501/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7501`

View PR using the GUI difftool: \
`$ git pr show -t 7501`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7501.diff">https://git.openjdk.java.net/jdk/pull/7501.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/util/regex/Pattern.java [hunk 1]
```diff
@@ -5091,11 +5097,18 @@ boolean match(Matcher matcher, int i, CharSequence seq) {
                             return false;
                     }
                 }
-                x += Character.charCount(c1);
+                int xIncr = Character.charCount(c1);
+                x += xIncr;
                 j += Character.charCount(c2);
+
+                if(xIncr > 1) {
```
Comment (naotoj) [inline src/java.base/share/classes/java/util/regex/Pattern.java:5104]: You can eliminate `xIncr` by comparing `c1 >= Character.MIN_SUPPLEMENTARY_CODE_POINT` here.

**Review by naotoj — COMMENTED** [review]
Looks good with a minor suggestion.

File: src/java.base/share/classes/java/util/regex/Pattern.java (hunk 1 above)
Comment (igraves) [reply src/java.base/share/classes/java/util/regex/Pattern.java:5104]: Nice! Thanks will do.

**Review by naotoj — APPROVED** [review]
Looks fine.

**Pull request closed without merging**